xds/src/main/java/io/grpc/xds/internal/certprovider/FileWatcherCertificateProvider.java

```diff
@@ -110,17 +81,18 @@ public void start() {
 
   @Override
   public void close() {
-    if (scheduledHandle != null) {
-      scheduledHandle.cancel();
-      scheduledHandle = null;
+    shutdown = true;
+    if (scheduledFuture != null) {
+      scheduledFuture.cancel(true);
+      scheduledFuture = null;
     }
     getWatcher().close();
   }
 
   private void scheduleNextRefreshCertificate(long delayInSeconds) {
-    RefreshCertificateTask runnable = new RefreshCertificateTask();
-    scheduledHandle =
-        syncContext.schedule(runnable, delayInSeconds, TimeUnit.SECONDS, scheduledExecutorService);
+    if (!shutdown) {
+      scheduledFuture = scheduledExecutorService.schedule(this, delayInSeconds, TimeUnit.SECONDS);
```

Comment:
In extreme condition when `delayInSeconds = 0`, the assignment of `scheduledFuture` may have a race because `checkAndReloadCertificates()` calls `scheduleNextRefreshCertificate()` again and makes another assignment of `scheduledFuture`.

Reply:
It sounds a little far fetched: `schedule` will add the current `Runnable` to a queue and return immediately. Even with 0 delay, `checkAndReloadCertificates()` has at least a couple of I/O operations which are blocking. So the return from the current `schedule` will happen before the  next call to `scheduleNextRefreshCertificate()` from ``checkAndReloadCertificates()`. In any case to address this I can check that delayInSeconds is > 0 in `FileWatcherCertificateProviderProvider` where the config is parsed.

Reply:
I have added a check in `FileWatcherCertificateProviderProvider` to not allow `refreshInterval <= 0`